File: medium/73. Set Matrix Zeroes.py
# 73. Set Matrix Zeroes
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.

        Use the first row and first col to store the states of each row and col.
        Note that the first row and the first col will use the same cell, the
        origin, so use that cell for the first row, and use a single var for
        the first col. Since we handle the first col separately, we need to
        skip the first col when iterating.
        Also note that when we use the states, we need to go from bottom to
        top, such that we don't erase our own footprint.

        Time: O(RC)
        Space: O(1)
        """
        firstColIsZero = False
        R, C = len(matrix), len(matrix[0])
        for r in range(R):
            if matrix[r][0] == 0:
                firstColIsZero = True
            for c in range(1, C):
                if matrix[r][c] == 0:
                    matrix[0][c] = matrix[r][0] = 0

        for r in reversed(range(R)):
            for c in range(1, C):
                if matrix[0][c] == 0 or matrix[r][0] == 0:
                    matrix[r][c] = 0
            if firstColIsZero:
                matrix[r][0] = 0

File: medium/73. Set Matrix Zeroes.rb
# 73. Set Matrix Zeroes
# @param {Integer[][]} matrix
# @return {Void} Do not return anything, modify matrix in-place instead.
def set_zeroes(matrix)
    """
    Note that in Ruby we can't let capitalized variables to be assigned when
    the method is called, because capitalized variables are CONSTANTS.

    Also note that to reversely iterate through a range, we can't use
    (bigger..smaller).each, but need to do (smaller..bigger).reverse_each.
    """
    rows = matrix.length
    cols = matrix[0].length
    is_first_col0 = false
    (0...rows).each do |r|
        is_first_col0 = true if matrix[r][0] == 0
        (1...cols).each do |c|
            if matrix[r][c] == 0
                matrix[r][0] = matrix[0][c] = 0
            end
        end
    end

    (0...rows).reverse_each do |r|
        (1...cols).each do |c|
            if matrix[0][c] == 0 or matrix[r][0] == 0
                matrix[r][c] = 0
            end
        end
        matrix[r][0] = 0 if is_first_col0
    end
end